clang/lib/Sema/SemaInit.cpp

```diff
@@ -727,6 +729,44 @@ void InitListChecker::FillInEmptyInitForField(unsigned Init, FieldDecl *Field,
     if (hadError || VerifyOnly) {
       // Do nothing
     } else if (Init < NumInits) {
```

Comment:
Is it OK that we don't warn on the `else` case here? Do we still warn for:

```c++
struct A { int x, y; };
A a = {.x = 1}; // missing `.y` initializer
```

Reply:
Yes, we still warn for this case.

Reply:
Hmm, how does that work? If this `if` condition doesn't hold and we created an `ImplicitValueInitExpr`, we wouldn't produce the warning.

Ah, I see what's happening: the `if` condition on line 736 is always true, because since cb77930d6b20e53c735233eecf4572a1c30eb0c0 we always resize the init list for a struct to be large enough for all fields, and our optimization to leave off trailing struct members was (accidentally, I think) removed. And we can see the optimization [stops working between Clang 3.6 and Clang 3.7](https://godbolt.org/z/fecM1K7Ec).

Given that we've been living without that optimization for 8 years and no-one seems to have noticed, I suppose that it's fine that we've lost it, but it looks like this function can be simplified a bit as a result. Half of the `if` condition on line 674 is always false, and this `if` condition is always true. Maybe consider changing the L674 condition to an `assert(Init < NumInits)` and simplify this too?

Reply:
> Maybe consider changing the L674 condition to an assert(Init < NumInits) and simplify this too?

I'm not sure I understand the suggestion. The condition on L674 looks like:
```
if (Init >= NumInits || !ILE->getInit(Init)) {
```
Even if I only remove `Init >= NumInits`, build of check-clang starts asserting with
```
Assertion `Init < getNumInits() && "Initializer access out of range!"'
```

Removing `!ILE->getInit(Init)` also doesn't seem logical to me since `FillInEmptyInitForField` is called for all fields in a record.

Reply:
Hm, right, we still call this with `Init >= NumInits` when in `VerifyOnly` mode (but *only* in that mode).

My concern here is that we're only warning in the `Init < NumInits` case, which is in principle wrong -- we should warn regardless of whether `Init < NumInits`. But in fact, the caller always resizes the `InitListExpr` so that `Init >= NumInits` whenever we're not in `VerifyOnly` mode, so I think this is probably not *actually* wrong, but it still *looks* wrong.

To make it not look wrong, I think we should remove all the dead code here: instead of checking whether `Init < NumInits` in non-`VerifyOnly` mode, we should assert that `Init < NumInits`, and delete the unreachable code that is trying to handle the impossible case: lines 684-685, 704-707, 772-778. Does that make sense?

Reply:
It seems `Init < NumInits` is always true for non-`VerifyOnly` mode only for structs. So, removing lines `684-685, 704-707, 772-778` breaks a number of tests with unions. The assertion on line 676 confirms:
```
if (Init >= NumInits || !ILE->getInit(Init)) {
  if (const RecordType *RType = ILE->getType()->getAs<RecordType>())
    if (!RType->getDecl()->isUnion())
      assert((Init < NumInits || VerifyOnly) &&
             "This ILE should have been expanded");
```
I can still put warning emission out of `Init < NumInits` check though if that seems more correct.